proxy/proxy/client.go

```diff
@@ -109,6 +109,27 @@ func (c *Client) Run(connSrc <-chan Conn) {
 }
 
 func (c *Client) handleConn(conn Conn) {
+	// Track connections count only if a maximum connections limit is set to avoid useless overhead
+	if c.MaxConnections > 0 {
+		// Acquire exclusive lock
+		connectionsCounterL.Lock()
+		if atomic.LoadUint64(&connectionsCounter) >= c.MaxConnections {
+			logging.Errorf("Too many open connections (max %d)", c.MaxConnections)
+			conn.Conn.Close()
+			connectionsCounterL.Unlock()
+			return
+		}
+
+		// Increment connectionsCounter then release exclusive lock
+		atomic.AddUint64(&connectionsCounter, 1)
```

Comment:
Don't use a lock plus atomics: you can use the return value from the AddUint64 to tell if this increment tipped over the counter.

    active := atomic.AddUint64(&connectionsCounter, 1)
    defer atomic.AddUint64(&connectionsCounter, ^uint64(0))
    if active > c.MaxConnections {
       conn.Conn.Close()
       logging.Errorf("Too many open connections (max %d)", c.MaxConnections)
    }